### PR TITLE
feat(vllm): vLLM ROCm backend via TheRock prebuilt

### PR DESCRIPTION
**Pull request by noamsto**
Refs #63. **Draft / WIP PoC** — packaging works end-to-end on NixOS and generates tokens on the GPU (gfx1150); still WIP for lemonade integration + gfx1151.

## Approach

Building vLLM + ROCm from source is blocked (nixpkgs `rocmPackages` is far behind ROCm 7.15 and lacks the Strix gfx targets), so this **repackages the upstream `lemonade-sdk/vllm-rocm` prebuilt** — a relocatable Python 3.14 prefix bundling torch + vLLM + a full TheRock ROCm for one gfx target, split across two <2 GB GitHub assets.

Lemonade already supports vLLM natively and resolves the backend via a `vllm.rocm_bin` config override, so no upstream patching of the launch path is needed — just point it at a nix-provided `vllm-server`.

## What works (validated on a gfx1150 / Strix Point host)

- `nix build .#vllm-rocm` succeeds (7.6 GB closure)
- `$out/bin/vllm-server --help` runs in a clean `env -i` (exit 0), printing the OpenAI API server usage with the `--model` / `--port` / `--max-model-len` args lemonade drives
- `import torch` → `2.11.0+rocm7.15.0` (`hip 7.15.0`); `import vllm` → `0.23.1.dev0`
- Module wiring behind `enableVllm` / `vllmGpuTarget`: `/etc/lemonade/backends/vllm-rocm` symlink + `vllm.rocm_bin` defaults seed + an `enableVllm requires enableROCm && enableLemonade` assertion

## The load-bearing packaging note

Do **not** `autoPatchelfHook` this bundle — it is self-contained (ships its own `rocm_sysdeps`, resolves `.so` via internal `$ORIGIN` runpaths). autoPatchelf rewrites those runpaths and injects nixpkgs gcc-15 `libstdc++`/`libatomic` over the Ubuntu-built torch, which **segfaults on `import torch`** (same class as #57). The working method is interpreter-only: set the nix ELF loader on `python3`/`clang`, keep every runpath, and supply the genuinely-missing system libs (`libstdc++`, `libz`) + the auditwheel `*.libs` dirs via the wrapper `LD_LIBRARY_PATH`.

## Not done yet (why this is a draft)

- [x] **Live GPU inference (standalone)** — validated on gfx1150: real tokens generated through ROCM_ATTN, 9.23 GiB GPU KV cache (see comment below). Still TODO: drive it *through lemonade*.
- [~] **gfx1151 (Strix Halo)** — hashes pinned + evaluates/builds; runtime validation pending a real Halo host
- [x] **README + CI** — enableVllm documented; module-eval-vllm check added
- [ ] Version-bump automation for the nightly-dated per-gfx tags (follow-up)

https://claude.ai/code/session_01Jxd7dRGeQvKTfd1cPEgG6j

### CODE REVIEW

**Comment by noamsto** [comment]
## ✅ Live GPU inference validated (gfx1150 / Radeon 890M)

Ran the built `$out/bin/vllm-server` standalone against a real model on this Strix Point host — full end-to-end generation on the GPU:

```
$ vllm-server --model Qwen/Qwen2.5-0.5B-Instruct --port 8009 \
    --enforce-eager --max-model-len 2048 --gpu-memory-utilization 0.4
...
[rocm.py:583]  Overriding with ROCM_ATTN out of potential backends: [ROCM_ATTN, TRITON_ATTN]
[gpu_worker.py:480]  Available KV cache memory: 9.23 GiB
[kv_cache_utils.py:1744]  GPU KV cache size: 806,384 tokens
[core.py:313]  init engine (profile, create kv cache, warmup model) took 6.10 s
INFO:     Application startup complete.

$ curl localhost:8009/v1/chat/completions -d '{"model":"test","messages":[{"role":"user","content":"..."}],"max_tokens":32}'
{"content":"vLLM on ROCm works! ... ROCm is designed for high-performance",
 "usage":{"prompt_tokens":40,"completion_tokens":32}}
```

So the relocated bundle does real GPU compute (ROCm attention backend, 9.23 GiB GPU KV cache), not just arg-parsing. No missing libs surfaced on the compute path, and Triton/ROCm init completed cleanly.

Remaining before un-drafting: end-to-end **through lemonade** (`vllm.rocm_bin` → backend `installed` → serve via the OpenAI API), **gfx1151** pin+validate on a Halo host, and the merge polish (CI eval check, README, version bump).

**Comment by noamsto** [comment]
## ✅ End-to-end through lemonade (gfx1150)

Wired `vllm.rocm_bin` → our nix `vllm-server` and served a model via lemonade's OpenAI API:

```
(Router) Model loaded successfully: Qwen3.5-0.8B-FP16-vLLM
(Server) POST /api/v1/chat/completions - 200 OK

$ curl .../api/v1/chat/completions -d '{"model":"Qwen3.5-0.8B-FP16-vLLM","messages":[{"role":"user","content":"what is NixOS?"}]}'
{"content":"NixOS is a Linux operating system designed to provide a consistent, reproducible,
 and secure environment...","usage":{"completion_tokens":27}}
```

**Two findings from the integration:**

1. **`global_timeout` bug (fixed in this PR).** vLLM alone passes `global_timeout` as its `wait_for_ready()` startup timeout (llama.cpp/whisper/sd use the built-in 600s default). Our `global_timeout=0` (which disables llama.cpp's 300s *request* cutoff) then means *0 startup attempts* → instant `vllm-server failed to start within timeout`. The module now sets a large finite `global_timeout` when `enableVllm`.

2. **The downloaded bundle confirms the premise of #63.** A pre-existing `~/.cache/lemonade/bin/vllm/rocm` bundle (python3.12, FHS) failed to load on NixOS ("Incompatible model file") — exactly the report. Our nix-repackaged `vllm-server` (python3.14) loads and serves in its place.

Remaining before un-drafting: gfx1151 pin+validate (prefetch in progress), and merge polish (CI eval check, README).

**Comment by rabejens** [comment]
Unfortunately I don’t have time this weekend but I will give it a shot on my Strix Halo machine soon. Thanks for quickly responding!
